grafana/grafana
# Plugins: Sandbox frontend plugins DOM access.

**What is this feature?**

Follow up to https://github.com/grafana/grafana/pull/68889 to introduce DOM protection for plugins running inside a frontend sandbox.

### General ideas behind the implemented protections:

- Plugins can only access DOM elements created and rendered by themselves.
- Plugins can only add DOM elements inside their own sandbox space (e.g. can't add elements to the document body directly). The exception here is the document.head where styles can be added.
- Plugins can't create or modify iframes in any possible way
- Plugins can't inject new scripts tags

It is much likely that this first implementation doesn't fully cover all the cases for the previously described points but rather than trying to be a full protection it tries to be a start point.

### Taken in consideration for this PR:

* Panel plugins using web workers, maps, canvas and displaying tooltips
* Datasource plugins query editors and configuration editors
* Apps full pages, navigation links and extensions

### Problematic case

Some plugins use third party libraries (specifically [tippyjs](https://github.com/atomiks/tippyjs-react/)) that rely on having full access to the document body to position tooltips.

Some plugins also make use of the body to render scaffolding elements (similar to a virtual DOM) to later move the content into the panel. 

Because of these cases we are initially allowing access to the document body but only dom elements that are not directly inside the grafana app unless they are inside the plugin sandbox.

**Why do we need this feature?**

Increase the security of the grafana platform.

**Who is this feature for?**

All plugins developers.

**Which issue(s) does this PR fix?**:

Closes: https://github.com/grafana/grafana/issues/69602
Tracked in: https://github.com/grafana/grafana/issues/68883


## How to test this?

Running the code from this branch:

* Install any type of non-angular non-inbuilt angular plugin from the catalog
* Use the plugin as usual
* The plugin should work as expected.
